[rebranch] Various fixes to get macos to a compiling state

bnbarham opened the pull request:
Haven't run tests at all, but this at least gets past all the compilation errors.

bnbarham committed:
[next] Remove LLVM_ATTRIBUTE_DEPRECATED

`LLVM_ATTRIBUTE_DEPRECATED` was removed in llvm/llvm-project
903c30f4d1f3bc0d1aae9ca83af17c0062d02b40. Use `[[deprecated]]` directly.

bnbarham committed:
[next] Use new VersionTuple API

The `VersionTuple` API was changed llvm/llvm-project
219672b8dd06c4765185fa3161c98437d49b4a1b to return `VersionTuple`
from `get*Version` rather than pass in major, minor, and subminor output
parameters. Update uses to the new API.

Note that `getMacOSXVersion` is slightly different in that it returns a
boolean while taking a `VersionTuple` output parameter to match its
previous behaviour. There doesn't seem to be any use that actually
checks this value though, so we should either update the API to return
an `Optional` and actually check it *or* remove the "failure" case and
return a `VersionTuple` like all the others.

bnbarham committed:
[next] Handle new LLVM triples

bnbarham committed:
[next] Add missing include

llvm/llvm-project 25991aadcc57fdada57147a9c83ca6301f6430bb removed
`llvm/Remarks/Remark.h` from `llvm/Remarks/RemarkParser.h`. Include it
directly.

bnbarham committed:
[next] Add default handling for new IITDescriptor::AnyPtrToElt

Added in llvm/llvm-project 49c2206b3bdce4a37a4602527b2d3da673514333.
Unhandled for now.

bnbarham committed:
[next] Add missing WithColor include

Best guess is that this was removed in a transitive include in
llvm/llvm-project 75e164f61d391979b4829bf2746a5d74b94e95f2.

bnbarham committed:
[next] Update dependency scanner to use new Clang API

apple/llvm-project 000a90ab63ce0595d442b3e7e8adc9c0506d2525 removed
`getAdditionalArgsWithoutModulePaths`. Based on the v2 libclang API, we
should be using `getCommandLineWithoutModulePaths` now.

bnbarham committed:
[next] Temporarily use getPointerElementType

`PointerType::getElementType` has been removed entirely as part of the
opaque pointers migration. Update to `getPointerElementType` for now
until we've also migrated.

bnbarham committed:
[next] Update ParseTopLevelDecl callers

`ParseTopLevelDecl` was changed in llvm/llvm-project
ab28488efe6de6f8fa856a1dfd8c0320d41d7608 and now takes a
Sema::ModuleImportState output parameter rather than the previous
(defaulted) boolean.

bnbarham committed:
[next] Update how supplementary outputs are removed

`ArgList::eraseArgIf` was removed in apple/llvm-project
000a90ab63ce0595d442b3e7e8adc9c0506d2525. Re-implement by retrieving the
option specifier ID for all options that have the `SupplementaryOutput`
flag in the given arguments and then erasing those.

bnbarham committed:
[next] Rename BitInt to ExtInt

This was renamed in llvm/llvm-project
6c75ab5f66b403f7ca67e86aeed3a58abe10570b.

bnbarham committed:
[next] Update `PPCallbacks::InclusionDirective` overrides

`InclusionDirective` was changed in llvm/llvm-project
d79ad2f1dbc2db63121620f55d6cfa915f2733ac to pass an
`Optional<FileEntryRef>` rather than a `FileEntry *`. Update overrides
in Swift to use the new API.

bnbarham committed:
[next] Define DEBUG_TYPE after includes

Not sure which LLVM commit added an `#undef` for `DEBUG_TYPE` in a header
transitively included by these files, but it doesn't seem uncommon
for headers to define and undefine their own `DEBUG_TYPE`.

Move `#define DEBUG_TYPE` after the includes to prevent it being
undefined.

bnbarham committed:
[next] Replace include for findDbgValues

`findDbgValues` was moved from `llvm/Transforms/Utils/Local.h` into
`llvm/IR/DebugInfo.h` in llvm/llvm-project
0ebf9a8e34b6aeee520aa5dc6916a155c5789533. Include that header instead.

bnbarham committed:
[next] Handle new ObjectFormatTypes

LLVM has added two new object format types, add them as unhandled for
now.

bnbarham committed:
[next] Add missing include

Not sure which header it was removed from. Could be worth trying IWYU at
some point.

bnbarham committed:
[next] Pass LLVM context to AttrBuilder constructor

The `AttrBuilder` API was changed in llvm/llvm-project
d2cc6c2d0c2f8a6e272110416a3fd579ed5a3ac1 to now take a `LLVMContext &`
as part of its constructor. Update uses to pass one in.

bnbarham committed:
[next] Handle new BTFTagAttributedType

Introduced in llvm/llvm-project
3251ba2d0fcf5223fce3e270b91c54f548664b4e.

bnbarham committed:
[next] Desugar new UsingType

llvm/llvm-project added a new `UsingType` for types found via
`UsingDecl` in e1600db19d6303f84b995acb9340459694e06ea9. Desugar in
clang importer to previous `RecordDecl`.

bnbarham commented:
@swift-ci please test

bnbarham commented on lib/ClangImporter/Serializability.cpp:
```diff
@@ -315,6 +315,10 @@ namespace {
     }
+
+    void writeBTFTypeTagAttr(const clang::BTFTypeTagAttr *attr) {
+      IsSerializable = false;
```
After talking to @rjmccall about this, I'll put up another PR after this one to instead handle `BTFTagAttributeType` in `DataStreamBasicWriter::writeQualType` by grabbing the underlying type and merging its qualifiers, then convert this to a fatal error since it should never occur. This is good enough to at least get us compiling though :)

bnbarham replied:
Went ahead and did this now since I was waiting on clang fixes anyway.

bnbarham commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -929,6 +929,8 @@ namespace {
     SUGAR_TYPE(Adjusted)
     SUGAR_TYPE(SubstTemplateTypeParm)
     SUGAR_TYPE(Elaborated)
+    SUGAR_TYPE(Using)
```
Not sure the best person to ask about this one - @zoecarver since it's C++ only?

bnbarham commented on lib/Driver/DarwinToolChains.cpp:
```diff
@@ -588,76 +589,62 @@ toolchains::Darwin::addDeploymentTargetArgs(ArgStringList &Arguments,
       switch (getDarwinPlatformKind((triple))) {
       case DarwinPlatformKind::MacOS:
-        triple.getMacOSXVersion(major, minor, micro);
+        triple.getMacOSXVersion(osVersion);
```
@DougGregor `getMacOSXVersion` can actually "fail". In that case `osVersion` is currently still set, but it'll return false. Do we want to handle that at all? I couldn't find a single case where we actually handle the error, so maybe it would be better to just modify the API to ignore the failing case?

bnbarham commented on lib/Basic/Platform.cpp:
```diff
@@ -181,11 +181,14 @@ StringRef swift::getPlatformNameForTriple(const llvm::Triple &triple) {
   case llvm::Triple::Ananas:
   case llvm::Triple::CloudABI:
   case llvm::Triple::DragonFly:
+  case llvm::Triple::DriverKit:
```
@egorzhdan I assume we do actually want to handle this?

egorzhdan replied:
Yes, I am planning to do that. I've started upstreaming DriverKit support in the compilers, so far only the LLVM part of it has landed. We would likely need the Clang part to land first before handling this.

bnbarham replied:
Thanks, so this is fine for now?

egorzhdan replied:
Yes, this is fine for now 👍

bnbarham commented:
apple/llvm-project#4276

@swift-ci please test

bnbarham commented:
https://github.com/apple/llvm-project/pull/4276

@swift-ci please test

bnbarham committed:
[next] Do not use libcxx standalone build

This was removed in llvm/llvm-project
6f17768e11480063f4c2bcbeea559505fee3ea19. Use new build instead.

bnbarham commented:
https://github.com/apple/llvm-project/pull/4276

@swift-ci please test

bnbarham commented:
@swift-ci please test macOS platform

bnbarham commented:
@swift-ci please test macOS platform

bnbarham commented:
https://github.com/apple/llvm-project/pull/4276

@swift-ci please test

bnbarham commented:
@swift-ci please test macOS platform

bnbarham commented:
Up to LLDB failures for macOS! Going to merge this set.
